src/Data/Text/Utf8.hs

```diff
@@ -211,6 +281,32 @@ decode4 :: CodeUnit -> CodeUnit -> CodeUnit -> CodeUnit -> Int
 decode4 cu0 cu1 cu2 cu3 =
   (fromIntegral cu0 .&. 0x7) `shiftL` 18 .|. (fromIntegral cu1 .&. 0x3f) `shiftL` 12 .|. (fromIntegral cu2 .&. 0x3f) `shiftL` 6 .|. (fromIntegral cu3 .&. 0x3f)
 
+indexCodePoint :: Text -> CodeUnitIndex -> (Int, CodePoint)
+indexCodePoint (Text u8data off _len) (CodeUnitIndex index) =
+  unsafeIndexCodePoint u8data $ CodeUnitIndex $ off + index
+
+-- | Decode a code point at the given 'CodeUnitIndex'.
+-- Returns garbage if there is no valid code point at that position.
+-- Does not perform bounds checking.
+-- See 'decode2', 'decode3' and 'decode4' for the expected format of multi-byte code points.
+unsafeIndexCodePoint :: ByteArray -> CodeUnitIndex -> (Int, CodePoint)
+{-# INLINE unsafeIndexCodePoint #-}
+unsafeIndexCodePoint !u8data (CodeUnitIndex !idx)
+  | cu0 < 0xc0 = (1, fromIntegral cu0)
+  | cu0 < 0xe0 = (2, decode2 cu0 (cuAt 1))
+  | cu0 < 0xf0 = (3, decode3 cu0 (cuAt 1) (cuAt 2))
+  | otherwise = (4, decode4 cu0 (cuAt 1) (cuAt 2) (cuAt 3))
+  where
+    cuAt !i = indexTextArray u8data $ idx + i
+    !cu0 = cuAt 0
+
+{-# INLINE unsafeIndexCodeUnit #-}
+-- | Get the code unit at the given index.
+unsafeIndexCodeUnit :: Text -> CodeUnitIndex -> CodeUnit
+unsafeIndexCodeUnit (Text u8data off len) (CodeUnitIndex index)
+  | index < 0 || index >= len = error $ "Index out of bounds " ++ show index
```

Comment:
How is it unsafe when there's bounds checking? Or is this a debugging left-over?

Reply:
You're right, this is a debugging remnant. I will clean up the naming scheme here a bit.